### ISSUE TITLE
imgformat -> webp

### ISSUE THREAD

**Issue by MarcDoughty** [overviewer/Minecraft-Overviewer#1868]
This is a minor housekeeping issue.

Every major browser in their current version, including long-term releases, supports webp images. I think switching to webp as a default is a good housekeeping task.

**Comment by CounterPillow** [comment]
webp is supported as an option for people to enable, but I'm strongly against making it the default. This is an opinion formed on the basis of much more than browser support. I'm currently quite busy so I cannot provide side-by-side comparisons here, but trust me that I've done the legwork in the past.


# Lossy WebP is not good quality

Back when Mozilla evaluated WebP by comparing it against their quality-optimised JPEG encoder in controlled trials, they've discovered that the benefit it provided was almost negligible. I very much doubt that they've changed their opinion on this since; rather, they've been forced to implement WebP due to their shrinking browser market share and the Chromium engine's dominance.

The core reason for this is that the current de-facto webp encoder is very PSNR biased. In layman's terms, it errs towards the side of being blurrier rather than being noisier. This makes it look good in metrics such as PSNR, but looks markedly worse to humans, as humans do not perceive the quality of images as a metric of peak signal to noise ratio.

Therein also lies the reason why WebP is a particularly bad fit for Overviewer: Overviewer's higher level tiles have lots of high frequency content. WebP absolutely destroys these with smudgy blurs, as you can verify yourself by looking at any forest:

![ouchwebp](https://user-images.githubusercontent.com/308818/98338050-d6794b80-2009-11eb-9c0e-efc5247a9ea5.png)

Since WebP is mainly a Google effort and Google has different opinions as to what constitutes pleasant artefacts than I do (judging by how absolutely horrid most YouTube videos look), I don't think we're going to see any strong psychovisual tuning in any WebP encoder any time soon.

More devastatingly, if you do not care about speed, lossy WebP clearly loses against quantised PNG for Overviewer's kind of output in terms of quality per filesize. If Overviewer had native support for image quantisation (instead of having to call out to pngnq for this) then it'd probably be even more clearly a bad idea to use lossy WebP.


# WebP has no future

Okay, so as you might already be pointing out, WebP has a lossless mode. And it does indeed perform better than PNG on Overviewer's tiles, and from a quick run, isn't even noticeably slower. However, I do not believe that WebP will stick around for long.

WebP was developed behind closed doors out of the VP8 video codec, also developed behind closed doors. VP8's goal was mainly to be comparable in quality to H.264 (quite an old codec by now!) while avoiding patent licensing issues. It failed to deliver in both aspects.

Google moved on to developing VP9, which did gain some traction thanks to the absolute mess that is HEVC patent licensing, and then began working on a VP10 which eventually was developed further in an open committee called the Alliance for Open Media (AOM) with many other companies, such as Mozilla, Microsoft, Cisco and so forth. This became the video coding standard now known as AV1, and has enjoyed its fair share of support from across the Web giants, with hardware decoders for it now starting to appear in consumer products, and multiple competing encoder projects being developed, all with their own little niche.

Why am I telling you all this? Because naturally, AV1 also was spun off into an image format, called AVIF. Modelled after HEIF, it uses the coding techniques pioneered by AV1 inside the same container as HEIF. Widespread support of AVIF is not there yet, but will be inevitable as the companies behind it finish laying the groundwork.

For WebP, this means it has become pretty much obsolete just as it gained support in more than just a few browsers. While PNG and JPEG have been around for donkey's ages, the same cannot be said about WebP. It seems to enjoy its widest popularity among automatic content optimisation done by news sites and such, and rarely does anyone actually save an original image as WebP and only deliver it as WebP.

So I would not be surprised if WebP suddenly disappeared within the next decade, without much fanfare. It's not like image formats can never be removed from the web once they've been supported by something; Mozilla axed MNG after all.


# Coming back to the initially dismissed support concern

While most major browsers do support WebP now (though Safari only does so in a tech preview, apparently), the same cannot be said of all Pillow installations we might reasonably be run with. WebP is, as far as I'm aware, an external dependency of Pillow, owing to the WebP format's complexity. More importantly, it's also an optional dependency. Making the default image output a format that is optional in our image writing dependency is not a good idea.

# Conclusion

If you want to use WebP, feel free to do so in your config. But there is a lot that speaks against it as a format, especially looking towards the future. That is why I am very insistent on it not being made the default.
